### DISCUSSION TITLE
DecRef (in Goroutine) leads to crash

### DISCUSSION THREAD

**Discussion by christian-korneck** [go-python/cpy3#16] [Usage Discussions]
originally posted by @jshiwam as comment to [#5](https://github.com/go-python/cpy3/pull/5#issuecomment-1155360816) .
moved from the PR to github discussions by the mod.

author: @jshiwam   
date: 14-Jun-22   

> hi team,
> 
> I have a program where I want to DecRef a PyObject(the return value of python methods) in a go routine. The code inside go routine looks something like this:
> 
```go
func MyFunc(request Request, requestBuilderFunc *python3.PyObject, pyTuple *python3.PyObject) *python3.PyObject
{
runtime.LockOSThread()
defer wg.Done()
defer runtime.GC()
defer debug.FreeOSMemory()
_gstate := python3.PyGILState_Ensure()

python3.PyTuple_SetItem(pyTuple, 0, python3.PyByteArray_FromStringAndSize(request.getRequestProtobytes()))
python3.PyTuple_SetItem(pyTuple, 1, python3.PyUnicode_FromString(request.getRequestType()))
python3.PyTuple_SetItem(pyTuple, 2, python3.PyUnicode_FromString(request.getRequestType()+"_"+strconv.Itoa(request.getRequestRCLID())))

// var resultant_dict *python3.PyObject
resultant_dict_from_proto := requestBuilderFunc.CallMethodArgs("from_proto_obj", pyTuple)
defer resultant_dict_from_proto.DecRef()

resultant_dict := requestBuilderFunc.CallMethodArgs("create_request_dict", resultant_dict_from_proto)
defer resultant_dict.DecRef()

resultant_dict_enriched := requestBuilderFunc.CallMethodArgs("enrich_request", resultant_dict)
defer resultant_dict_enriched.DecRef()

requestBuilderFunc.CallMethodArgs("create_request_temp_table", resultant_dict_enriched)
python3.PyGILState_Release(_gstate)

fmt.Println("Request Builder Ends")

return resultant_dict_enriched
}
```
> 
> But this code crashes with following stack trace:
> 
> runtime stack:
> runtime.throw(0xb369f9, 0x2a)
> /u1/tags/dev/devtools/golang/1.16.15/src/runtime/panic.go:1117 +0x72
> runtime.sigpanic()
> /u1/tags/dev/devtools/golang/1.16.15/src/runtime/signal_unix.go:718 +0x2e5
> 
> goroutine 82 [syscall, locked to thread]:
> runtime.cgocall(0x9d4db0, 0xc000125e20, 0xc0009d4db0)
> /u1/tags/dev/devtools/golang/1.16.15/src/runtime/cgocall.go:154 +0x5b fp=0xc000125df0 sp=0xc000125db8 pc=0x5a4c9b
> github.com/go-python/cpy3._Cfunc_Py_DecRef(0x7f8efb18ad70)
> _cgo_gotypes.go:5135 +0x3c fp=0xc000125e20 sp=0xc000125df0 pc=0x6a077c
> github.com/go-python/cpy3.(*PyObject).DecRef.func1(0x7f8efb18ad70)
> /u1/shijaisw/go/pkg/mod/github.com/go-python/cpy3@v0.0.0-20220603234950-407b83a936e2/object.go:48 +0x4d fp=0xc000125e50 sp=0xc000125e20 pc=0x6a2c0d
> github.com/go-python/cpy3.(*PyObject).DecRef(0x7f8efb18ad70)
> /u1/shijaisw/go/pkg/mod/github.com/go-python/cpy3@v0.0.0-20220603234950-407b83a936e2/object.go:48 +0x2b fp=0xc000125e68 sp=0xc000125e50 pc=0x6a130b
> github.com/jshiwam/memolk/libqcaas.requestBuilder(0xbbfc60, 0xc0002ec580, 0x7f8f282daf90, 0x7f8efb321960, 0x7f8efb18ad70)
> /u1/shijaisw/DQE/memolk/libqcaas/requestProcessor.go:97 +0x63b fp=0xc000125fb8 sp=0xc000125e68 pc=0x8d593b
> runtime.goexit()
> /u1/tags/dev/devtools/golang/1.16.15/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc000125fc0 sp=0xc000125fb8 pc=0x60cba1
> created by github.com/jshiwam/memolk/libqcaas.RequestProcessor
> /u1/shijaisw/DQE/memolk/libqcaas/requestProcessor.go:39 +0x373
> 
> goroutine 1 [semacquire]:
> sync.runtime_Semacquire(0x1288ce8)
> /u1/tags/dev/devtools/golang/1.16.15/src/runtime/sema.go:56 +0x45
> sync.(*WaitGroup).Wait(0x1288ce0)
> /u1/tags/dev/devtools/golang/1.16.15/src/sync/waitgroup.go:130 +0x65
> github.com/jshiwam/memolk/libqcaas.RequestProcessor(0xb1f09f, 0x8, 0xbbfc60, 0xc0002ec580, 0xc000524ae0)
> /u1/shijaisw/DQE/memolk/libqcaas/requestProcessor.go:42 +0x396
> github.com/jshiwam/memolk/libqcaas.ProcessRequest(0xb1f09f, 0x8, 0xbbfc60, 0xc0002ec580, 0xc000524ae0, 0xc00051c620)
> /u1/shijaisw/DQE/memolk/libqcaas/requestProcessor.go:20 +0xc5
> main.RunQCaaS(0xc000524ae0)
> /u1/shijaisw/DQE/memolk/reqSender.go:96 +0x84c
> main.main()
> /u1/shijaisw/DQE/memolk/main.go:39 +0x170
> 
> goroutine 8 [select]:
> database/sql.(*DB).connectionOpener(0xc0001c56c0, 0xbb8eb8, 0xc000371000)
> /u1/tags/dev/devtools/golang/1.16.15/src/database/sql/sql.go:1133 +0xb5
> created by database/sql.OpenDB
> /u1/tags/dev/devtools/golang/1.16.15/src/database/sql/sql.go:740 +0x12a

**Comment by christian-korneck** [comment]
@jshiwam one of the first things I notice is there's no proper error handling, nil checks. This is how I tend do it:
https://github.com/christian-korneck/python-go/blob/b9bf06c9e65d60d1026dceb47dcdf817d11fc68e/py-bindings/outliers/main.go#L55

```go
detectdataPy := detect.CallObject(args)
	if !(detectdataPy != nil && python3.PyErr_Occurred() == nil) { //retval: New reference
		python3.PyErr_Print()
		return nil, fmt.Errorf("error calling function detect")
	}
defer detectdataPy.DecRef()
```

(note that this might not necessarily be the best way to do it, just what I tend to use)

**Reply by jshiwam** [reply]
Hi @christian-korneck , thanks for the error handling part, the main issue is something else here. The reason behind the above code crashing while I try to DecRef the PyObject(pyTuple  argument in the code that I posted) is as follows:

```go
state := python3.PyEval_SaveThread() // release the GIL, the GIL is unlocked for using by goroutines
	// Restore the state and lock the GIL
	defer python3.PyEval_RestoreThread(state) // acquire the GIL, the GIL is locked by the main thread
	args := python3.PyTuple_New(1)            // new reference, a call DecRef() is needed
	if args == nil && python3.PyErr_Occurred() != nil {
		err = errors.New("error creating python tuple object")
		return
	}
	defer args.DecRef()
	defer fmt.Println("DecReffing Args", args)

	ret := python3.PyTuple_SetItem(args, 0, limit) // steals reference to limit

	// Cleans the Go variable, because now a new owner is caring about related PyObject
	// no action, such as a call DecRef(), is needed here
	limit = nil

	if ret != 0 {
		err = errors.New("error setting a tuple item")
		limit.DecRef()
		limit = nil
		return
	}

	// for true {
	wg.Add(20)

	for i := 0; i < 10; i++ {
		go func() {
			runtime.LockOSThread()
			_gstate := python3.PyGILState_Ensure() // acquire the GIL, the GIL is locked by the 1st goroutine
			odds.Call(args, python3.PyDict_New())
			python3.PyGILState_Release(_gstate) // release the GIL, the GIL is unlocked for using by others

			wg.Done()
		}()

		go func() {
			runtime.LockOSThread()
			_gstate := python3.PyGILState_Ensure() // acquire the GIL, the GIL is locked by the 2nd goroutine
			even.Call(args, python3.PyDict_New())
			python3.PyGILState_Release(_gstate) // release the GIL, the GIL is unlocked for using by others

			wg.Done()
		}()
	}
	wg.Wait()
	// }
```

Here I have used the code in @esendjer's PR, he probably needs to add `runtime.LockOSThread()` at the start of each go routine, anyways the problem here is that I have DecRefed the args before Locking the GIL. It gives me segfault. But when I print the args PyObject struct before decreffing it  I get this 
`DecReffing Args &{1 0x10061e0}`

As you can see the [PyObject Ref Count](https://github.com/python/cpython/blob/f9433fff476aa13af9cb314fcc6962055faa4085/Include/object.h#L41) here is 1. I don't understand why I can't decref a PyObject when the GIL is in unlocked state. Are there any refs that I can go through to understand it? 

Thanks!

**Reply by jshiwam** [reply]
```
Save the state and lock the GIL.
Do Python.
Restore the state and unlock the GIL.
```

I have gone through this btw, I am confused on whether this needs to be done in the main thread as well? Because there is no concurrency at the level of main there, there is only one main thread here.

**Reply by christian-korneck** [reply]
@jshiwam I don't really get your point. Can you post a complete, runnable example?

Side note: I noticed that you've copied at least one "mistake" from the (unreviewed) PR code:
```
args := python3.PyTuple_New(1)            // new reference, a call DecRef() is needed
	if args == nil && python3.PyErr_Occurred() != nil {
```

should probably not check for `python3.PyErr_Occurred()` as the docs does not mention that `PyTuple_New()` raises a Python error on failure. So this error handling probably doesn't work. There might be more.

**Reply by jshiwam** [reply]
Here 
https://go.dev/play/p/bKSuJVtEPv0

the above code will fail because there is a DecRef on line number 69 which is DecReffing a PyObject when the GIL is unlocked. My misunderstanding in this code was that, we only need to lock the gil and do python inside go routines, which is not true. Even if it is the main thread we still need to lock the gil before performing python operations.

**Reply by christian-korneck** [reply]
thanks, I’ll have a look.

**Comment by christian-korneck** [comment]
I think part of the problem is that in this example you're releasing the GIL in the `main()` function. Which means that you can't continue to use the GIL (= do python stuff) in the main function beyond this point (unless you re-aquire it).

```
	// Save the current state and release the GIL
	// so that goroutines can acquire it
	state := python3.PyEval_SaveThread() // release the GIL, the GIL is unlocked for using by goroutines
	// Restore the state and lock the GIL
	defer python3.PyEval_RestoreThread(state) // acquire the GIL, the GIL is locked by the main thread
```

However, directly after these lines Python stuff is happening in the main function. If you move the above block to the line before you start the Goroutines it doesn't seem to crash:

```
	// Save the current state and release the GIL
	// so that goroutines can acquire it
	state := python3.PyEval_SaveThread() // release the GIL, the GIL is unlocked for using by goroutines
	// Restore the state and lock the GIL
	defer python3.PyEval_RestoreThread(state) // acquire the GIL, the GIL is locked by the main thread

	// for true {
	wg.Add(20)

	for i := 0; i < 10; i++ {
		go func() {
...
```
https://go.dev/play/p/ogdAmVyl5F0

⚠️ Disclaimer (also for anyone else who might be reading this): I am suspicious that there might be more mistakes in the code, please don't regard the above code and link as a "good" example.

Maybe as a tiny bit of advice (maybe it helps you, maybe not). I highly recommend to look up _every_ Python C-API function that you call in the Python C-API docs. Imho it's the only way to avoid deadly surprises.
